zk/stages/stage_batches.go

```diff
@@ -900,3 +1041,97 @@ func writeL2Block(eriDb ErigonDb, hermezDb HermezDb, l2Block *types.FullL2Block,
 
 	return nil
 }
+
+// findCommonAncestor searches the latest common ancestor block number and hash between the data stream and the local db.
+// The common ancestor block is the one that matches both l2 block hash and batch number.
+func findCommonAncestor(
+	db erigon_db.ReadOnlyErigonDb,
+	hermezDb state.ReadOnlyHermezDb,
+	dsClient DatastreamClient,
+	latestBlockNum uint64) (uint64, common.Hash, error) {
+	var (
+		startBlockNum = uint64(0)
+		endBlockNum   = latestBlockNum
+		blockNumber   *uint64
+		blockHash     common.Hash
+	)
+
+	if latestBlockNum == 0 {
+		return 0, emptyHash, ErrFailedToFindCommonAncestor
+	}
+
+	for startBlockNum <= endBlockNum {
+		if endBlockNum == 0 {
+			return 0, emptyHash, ErrFailedToFindCommonAncestor
+		}
+
+		midBlockNum := (startBlockNum + endBlockNum) / 2
+		midBlockDataStream, err := dsClient.GetL2BlockByNumber(midBlockNum)
+		if err != nil &&
+			// the required block might not be in the data stream, so ignore that error
+			!strings.Contains(err.Error(), client.ErrBadFromBookmarkStr) {
```

Comment:
I wouldn't be worried about error string being changed, because I am using the string from the datastreamer library, by indexing the map of the error codes (https://github.com/0xPolygonHermez/cdk-erigon/blob/32b3a2926d9b86a9211caff1724d73cc048bc68d/zk/datastream/client/stream_client.go#L38):
```go
	ErrBadFromBookmarkStr = datastreamer.StrCommandErrors[datastreamer.CmdErrBadFromBookmark]
```
In case error string for the `bad from bookmark` gets changed, it is fine, because I'm retrieving the error string that corresponds to the `CmdErrBadFromBookmark` error code.

Your proposal would imply that `GetL2BlockByNumber` needs to return a `ResultEntry` instance (alongside with `error`) and to me it would be more cumbersome for handling from the caller perspective. WDYT?

Reply:
Understood the stream is imported from the library, but the problem might occur if the data stream client and the server are using different versions of data stream library. We can't be sure that, at any point of time, the versions of all data stream library used by all clients and servers are always in sync. 
My proposal is to add the error code returned by the server to the error string (if it isn't already there), and check the error code directly.

Reply:
Fixed https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/a21e94158cbc17039e153c96c1e087e52a4d159c.